# Example of Retrieval Evaluation using Contexts and Ground Truth Contexts

**Authors**: 
- Novan Parmonangan Simanjuntak (novan.p.simanjuntak@gdplabs.id)


**Reviewers**: 
- Komang Elang Surya Prawira (komang.e.s.prawira@gdplabs.id)
- Surya Mahadi (made.r.s.mahadi@gdplabs.id))

## References
[1] [GLAIR GenAI Internal SDK - Evaluate Relevance Between Retrieved Contexts and Ground Truth Contexts](TODO) \
[2] https://github.com/beir-cellar/beir

## Prepare Environment
Before we start, please make sure to install the SDK library.

To install the SDK library, you need to create a personal access token on GitHub. Please follow these steps:
1. You need to log in to your [GitHub Account](https://github.com/).
2. Go to the [Personal Access Tokens](https://github.com/settings/tokens) page.
3. If you haven't created a Personal Access Tokens yet, you can generate one.
4. When generating a new token, make sure that you have checked the `repo` option to grant access to private repositories.
5. Now, you can copy the new token that you have generated and paste it into the script below.

In [1]:
import getpass
import subprocess
import sys

def install_sdk_library():
    token = getpass.getpass("Input Your Personal Access Token: ")

    cmd = f"pip install git+https://{token}:x-oauth-basic@github.com/GDP-ADMIN/gen-ai-internal@f/retrieval_evaluator#egg=gdplabs_gen_ai[eval]"

    with subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True) as process:
        stdout, stderr = process.communicate()

        if process.returncode != 0:
            sys.stdout.write(stderr)
            raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
        else:
            sys.stdout.write(stdout)

install_sdk_library()

  Cloning https://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:****@github.com/GDP-ADMIN/gen-ai-internal (to revision f/retrieval_evaluator) to /tmp/pip-install-cgsy6yio/gdplabs-gen-ai_6b2489fcf2594a55abbfcc409baebc8b
  Resolved https://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:****@github.com/GDP-ADMIN/gen-ai-internal to commit b61b104922ffe2c860c6a1079c1d9660487c8c8a
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'


<b>Warning:</b>
After running the command above, you need to restart the runtime in Google Colab for the changes to take effect. Not doing so might lead to the newly installed libraries not being recognized.

To restart the runtime in Google Colab:
- Click on the `Runtime` menu.
- Select `Restart runtime`.

<b>Note: We highly recommend that you run the following code on a GPU instead of a CPU for optimal performance.</b>

Simply go to `Runtime` > `Change runtime type` in the menu, select the desired GPU as the `Hardware accelerator`, and hit `Save`. This ensures faster execution of the code. To fine-tune a Large Language Model (LLM) with 7 billion parameters or more without quantization, you will need a minimum of 40 GB of GPU memory, or an equivalent to A100 instances in Google Colab.

## Evaluate Contexts and Ground Truth Contexts
Once you have completed the previous step, you can start by importing the necessary library:

In [2]:
from typing import Dict, List
from gdplabs_gen_ai.evaluation import EvaluateRetrieval

Then prepare your dataset

In [3]:
# Ground truth contexts for each query with relevance scores. 
# Use query_id and context_id to represent query and context.
ground_truth_contexts: Dict[str, Dict[str, int]] = {
    "query1": {"doc1": 1, "doc3": 1, "doc5": 1},
    "query2": {"doc2": 1, "doc4": 1},
    "query3": {"doc6": 1, "doc8": 1, "doc10": 1},
    "query4": {"doc7": 1, "doc9": 1}
}

# Retrieved contexts for each query with similarity scores. It doesn't need to be sorted by scores.
# In this examples we retrieve top 5 contexts for each query.
retrieved_contexts: Dict[str, Dict[str, float]] = {
    "query1": {"doc1": 0.3, "doc2": 0.9, "doc3": 0.8, "doc4": 0.2, "doc5": 0.7},
    "query2": {"doc1": 0.8, "doc2": 0.3, "doc3": 0.6, "doc4": 0.2, "doc5": 0.4},
    "query3": {"doc6": 0.5, "doc7": 0.3, "doc8": 0.7, "doc9": 0.2, "doc10": 0.6},
    "query4": {"doc1": 0.8, "doc3": 0.8, "doc5": 0.5, "doc7": 0.3, "doc9": 0.2}
}

Prepare list of k values that you want to evaluate (NDCG@k, MAP@k, Recall@k, Precision@k)

In [4]:
# List of k values for evaluation
k_values: List[int] = [1, 3, 5]

After we prepare the dataset and k values, we can start the evaluation

In [5]:
# Evaluate retrieval performance using NDCG@k, MAP@K, Recall@K, and Precision@K
ndcg, map_score, recall, precision = EvaluateRetrieval.evaluate(
    ground_truth_contexts, retrieved_contexts, k_values
)

# Display the evaluation metrics
print(f"NDCG: {ndcg}")
print(f"MAP: {map_score}")
print(f"Recall: {recall}")
print(f"Precision: {precision}")


NDCG: {'NDCG@1': 0.25, 'NDCG@3': 0.38268, 'NDCG@5': 0.68384}
MAP: {'MAP@1': 0.08333, 'MAP@3': 0.34722, 'MAP@5': 0.57222}
Recall: {'Recall@1': 0.08333, 'Recall@3': 0.41667, 'Recall@5': 1.0}
Precision: {'P@1': 0.25, 'P@3': 0.41667, 'P@5': 0.5}


## Analyze results

The output provides key insights into the performance of the retrieval system. Let's break down each metric and interpret what it signifies:  

Normalized Discounted Cumulative Gain (NDCG):  
NDCG@1: 0.25  
NDCG@3: 0.38268  
NDCG@5: 0.68384  
Interpretation: NDCG measures the quality of the ranked retrieved documents based on their relevance. The values indicate that the relevance improves as more documents are considered (from 1 to 5). A score of 0.68384 at NDCG@5 suggests that the top 5 documents have a moderately high relevance. However, the lower score at NDCG@1 indicates that the most relevant document is not always ranked highest.  


Mean Average Precision (MAP):  
MAP@1: 0.08333  
MAP@3: 0.34722  
MAP@5: 0.57222  
Interpretation: MAP assesses the precision across all queries. The scores reflect a similar trend as NDCG, where precision improves with more documents. The low score at MAP@1 implies that the top-ranked document is often not the most relevant. However, considering more documents (up to 5) increases the likelihood of including relevant documents.  


Recall:  
Recall@1: 0.08333  
Recall@3: 0.41667  
Recall@5: 1.0  
Interpretation: Recall measures how many relevant documents are retrieved. The perfect score at Recall@5 indicates that all relevant documents are included within the top 5. However, the lower scores at Recall@1 and Recall@3 suggest that not all relevant documents are ranked at the very top.
Precision:  
P@1: 0.25  
P@3: 0.41667  
P@5: 0.5  
Interpretation: Precision evaluates how many of the retrieved documents are relevant. The scores show that while half of the top 5 documents are relevant (P@5), the precision at the very top (P@1) is lower.  

## Action based on metrics:  


1. Focus on Improving Top-Ranked Results: The lower scores at NDCG@1 and MAP@1 suggest a need to enhance the algorithm to ensure the most relevant documents are ranked higher.  
2. Balance Between Precision and Recall: While recall at 5 is perfect, precision is only 50%, indicating a trade-off. Aim to improve precision without significantly compromising recall.  
3. Analyze Individual Queries: Investigate queries where the retrieval system underperforms, especially those contributing to lower scores at lower 'k' values.  
4. Refine Retrieval Algorithms: Consider adjusting or fine-tuning the retrieval algorithm, possibly integrating more sophisticated techniques like semantic search or machine learning models that can better understand query intent.  
5. Iterative Testing and Evaluation: Continuously evaluate and refine the system using different datasets and queries to ensure robustness and versatility.  
